contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+- `BatchExportRun`: A `BatchExport` for each triggered workflow. Users aren’t meant to create these, only Temporal. They mostly serve to report on individual workflow executions.
+
+Users create a `BatchExport` that includes a schedule (which validates against the `BatchExportSchedule`) and a destination (stored as a `BatchExportDestination`). Data then gets read from PostHog and processed to write to the destination.
+
+With this structure in place, we began working on improving reliability, transparency, and performance.
+
+### Improving reliability and adding retries
+
+We prioritized improving reliability. As mentioned, Temporal has support for retry and timeout logic per activity and workflow which made a big difference. What we needed to do was tune this to our system and destinations. We needed to tune four main pieces:
+
+1. **ClickHouse** - [spaced out retries](https://github.com/PostHog/posthog/pull/15558) when reconnecting to prevent failures if in use by another query, retry [connection errors](https://github.com/PostHog/posthog/pull/16574) which are likely network failures rather than outages, use [offline ClickHouse cluster](https://github.com/PostHog/posthog/pull/16470) to improve performance and allowable query length, and sacrificed some performance to [de-duplicate batches](https://github.com/PostHog/posthog/pull/16354) by adding `DISTINCT ON` clause (because it handles queries asynchronously so de-duplication isn’t guaranteed).
+2. **Postgres** - [health checks](https://github.com/PostHog/posthog/pull/16079) to ensure it is healthy before running workflows and [retry forever](https://github.com/PostHog/posthog/pull/16703) because we control everything related to those activities so failures are transient (other than bugs).
+3. **Destinations** - define which destination errors are [retriable and which aren’t](https://github.com/PostHog/posthog/pull/16814), [reset the iterator when a `JSONDecodeError` happens](https://github.com/PostHog/posthog/pull/16473) to keep the workflow going, and handle [query output](https://github.com/PostHog/posthog/pull/16247) errors from Snowflake.
+4. **Temporal** -  add [max retries and errors](https://github.com/PostHog/posthog/pull/15399) so it doesn’t retry forever, add [heartbeat API to track activity](https://github.com/PostHog/posthog/pull/16494) because we lose state on deployment causing exports to reset, and create an [endpoint](https://github.com/PostHog/posthog/pull/16181) for resetting a `BatchExportRun`.
+
+### Making exports transparent (internally and externally)
+
+The second part of the move to Temporal was making processes  transparent both internally and externally.
+
+As mentioned in the last section, Temporal creates transparency into the errors workflows encounter. We can now also track types and records exported. Both these enable us to improve the export system further and faster than the old system.
```

Comment:
I think this is a typo:

```suggestion
As mentioned in the last section, Temporal creates transparency into the errors workflows encounter. We can now also track bytes and records exported. Both these enable us to improve the export system further and faster than the old system.
```

However, the old system also tracked these, so it may be a stretch to say we improved it further.

Reply:
Reworded slightly